### DISCUSSION TITLE
`HashMap` with `valuable` is represented as a string in when using JSON formatter

### DISCUSSION THREAD

**Discussion by banool** [tokio-rs/tracing#3039] [Q&A] [answered]
Say I have code like this:
```rust
fn my_function() {
    let mut mymap: HashMap<String, String> = HashMap::new();
    mymap.insert("hey".to_string(), "friend".to_string());
    
    info!(mymap = tracing::field::valuable(&mymap), blah=2);
}
```

I'm using `tracing_subscriber::fmt::layer().json()`.

The output looks like this:
```json
{
  "timestamp": "2024-07-24T09:04:17.987349Z",
  "level": "INFO",
  "target": "aptos_api_gateway::http_proxy::http_handler",
  "mymap": "{\"hey\": \"friend\"}",
  "blah": 2
}
```

I was hoping that the HashMap would be represented as JSON within the top level JSON, but as you can see, it's just a string representation.

Is there any way to get the output to look like this?
```json
{
  "timestamp": "2024-07-24T09:04:17.987349Z",
  "level": "INFO",
  "target": "aptos_api_gateway::http_proxy::http_handler",
  "mymap": {
    "hey": "friend"
  },
  "blah": 2
}
```

I have tried this instead:
```rust
#[derive(serde::Deserialize, serde::Serialize, valuable::Valuable)]
struct MyWrapper {
    mymap: HashMap<String, String>,
}

...

info!(mywrapper = tracing::field::valuable(&MyHeaders::new(headers)));
```

But I just get this:
```
"mywrapper":"MyWrapper { mymap: {\"hey\": \"friend\"} }"
```

**Comment by banool** [comment]
Linking back to the Discord thread: https://discord.com/channels/500028886025895936/819261130861051986/1265600592693166177.

**Comment by mladedav** [comment]
Are you using the [unstable features](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/#unstable-features) and the `valuable` feature on `tracing-subscriber`?

**Comment by davidbarsky** [comment, marked as answer]
As @mladedav said, you are likely not enabling the unstable valuable feature on `tracing-subscriber`.

**Comment by banool** [comment]
Ah yes turns out that was it! I had the unstable feature enable but not the `valuable` feature on `tracing-subscriber`. Thanks all!
